lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java

```diff
@@ -0,0 +1,177 @@
+    this.numHits = numHits;
+    this.totalHits = 0;
+  }
+
+  // We always return COMPLETE since this collector should ideally
+  // be used only with large number of hits case
+  @Override
+  public ScoreMode scoreMode() {
+    return ScoreMode.COMPLETE;
+  }
+
+  @Override
+  public LeafCollector getLeafCollector(LeafReaderContext context) {
+    final int docBase = context.docBase;
+    return new TopScoreDocCollector.ScorerLeafCollector() {
+
+      @Override
+      public void setScorer(Scorable scorer) throws IOException {
+        super.setScorer(scorer);
+        updateMinCompetitiveScore(scorer);
+      }
+
+      @Override
+      public void collect(int doc) throws IOException {
+        float score = scorer.score();
+
+        // This collector relies on the fact that scorers produce positive values:
+        assert score >= 0; // NOTE: false for NaN
+
+        if (totalHits < numHits) {
+          hits.add(new ScoreDoc(doc, score));
+          totalHits++;
+          return;
+        } else if (totalHits == numHits) {
+          // Convert the list to a priority queue
+
+          // We should get here only when priority queue
+          // has been built
+          assert pq == null;
+          assert pqTop == null;
+          pq = new HitQueue(numHits, false);
+
+          for (int i = 0; i < hits.size(); i++) {
+            pq.add(hits.get(i));
+          }
+
+          pqTop = pq.top();
+        }
+
+        if (score > pqTop.score) {
+          pqTop.doc = doc + docBase;
+          pqTop.score = score;
+          pqTop = pq.updateTop();
+          updateMinCompetitiveScore(scorer);
+        }
+        ++totalHits;
+      }
+    };
+  }
+
+  protected void updateMinCompetitiveScore(Scorable scorer) throws IOException {
+    if (pqTop != null) {
+      scorer.setMinCompetitiveScore(Math.nextUp(pqTop.score));
+    }
+    totalHitsRelation = TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO;
+  }
+
+  public TopDocs topDocs(int howMany) {
+
+    if (howMany <= 0 || howMany > totalHits) {
+      throw new IllegalArgumentException("Incorrect number of hits requested");
+    }
+
+    ScoreDoc[] results = new ScoreDoc[howMany];
+
+    // Get the requested results from either hits list or PQ
+    populateResults(results, howMany);
+
+    return newTopDocs(results);
+  }
+
+  /**
+   * Populates the results array with the ScoreDoc instances. This can be
+   * overridden in case a different ScoreDoc type should be returned.
+   */
+  protected void populateResults(ScoreDoc[] results, int howMany) {
+    if (pq != null) {
+      assert totalHits >= numHits;
+      for (int i = howMany - 1; i >= 0; i--) {
+        results[i] = pq.pop();
+      }
+      return;
+    }
+
+    // Total number of hits collected were less than numHits
+    assert totalHits < numHits;
+    Collections.sort(hits, new Comparator<ScoreDoc>() {
+      @Override
+      public int compare(ScoreDoc o1, ScoreDoc o2) {
+        if (o1.score < o2.score) {
+          return -1;
```

Comment:
This seems to be putting lower scores first, while we'd like the higher scores first?

Reply:
Removed the comparator